Libraries/PeriphDrivers/Include/MAX32655/mxc_sys.h

```diff
@@ -191,6 +191,91 @@ typedef enum {
 
 /***** Function Prototypes *****/
 
+typedef struct {
+    int ie_status;
+    int in_critical;
+} mxc_crit_state_t;
+
+static mxc_crit_state_t _state = { .ie_status = 0xFFFFFFFF, .in_critical = 0 };
+
+static inline void _mxc_crit_get_state()
+{
+#ifdef __CORTEX_M
+    /*
+        On ARM M the 0th bit of the Priority Mask register indicates
+        whether interrupts are enabled or not.
+
+        0 = enabled
+        1 = disabled
+    */
+    uint32_t primask = __get_PRIMASK();
+    _state.ie_status = (primask == 0);
+#endif
+#ifdef __riscv
+    /*
+        On RISC-V bit position 3 (Machine Interrupt Enable) of the
+        mstatus register indicates whether interrupts are enabled.
+
+        0 = disabled
+        1 = enabled
+    */
+    uint32_t mstatus = get_mstatus();
+    _state.ie_status = ((mstatus & (1 << 3)) != 0);
+#endif
+}
+
+/**
+ * @brief Enter a critical section of code that cannot be interrupted.
+ */
+static inline void MXC_SYS_Crit_Enter(void)
+{
+    _mxc_crit_get_state();
+    if (_state.ie_status)
+        __disable_irq();
+    _state.in_critical = 1;
+}
+
+/**
+ * @brief Exit a critical section of code, re-enabling interrupts if they
+ *        were previously.
+ */
+static inline void MXC_SYS_Crit_Exit(void)
```

Comment:
Why these are inline functions? Not saying it's wrong or needs to be changed, just curious.

Reply:
Inline is used to give the lowest possible timing overhead.  Since this API is essentially just a wrapper around an assembly function, it doesn't make sense to add the 2-4uS overhead associated with a normal function call.  Especially because the code needs to enter the critical section _as soon as possible_ to avoid getting interrupted on its way there.

Reply:
It's "static inline" as well because that is a forced requirement for accessing other static member variables and using static function calls.